libgnucash/app-utils/gnc-option-impl.cpp

```diff
@@ -0,0 +1,500 @@
+}
+
+QofInstance*
+qof_instance_from_string(const std::string& str, GncOptionUIType type)
+{
+    QofInstance* retval{nullptr};
+    // Commodities are often serialized as Namespace::Mnemonic or just Mnemonic
+    if (type == GncOptionUIType::CURRENCY ||
+        type == GncOptionUIType::COMMODITY)
+    {
+        auto book{gnc_get_current_book()};
+        auto table = gnc_commodity_table_get_table(book);
+        auto sep{str.find(":")};
+        if (sep != std::string::npos)
+        {
+            auto name_space{str.substr(0, sep)};
+            auto mnemonic{str.substr(sep + 1, -1)};
+            retval = QOF_INSTANCE(gnc_commodity_table_lookup(table,
+                                                             name_space.c_str(),
+                                                             mnemonic.c_str()));
+        }
+        if (!retval && type == GncOptionUIType::CURRENCY)
+            retval = QOF_INSTANCE(gnc_commodity_table_lookup(table,
+                                                             "CURRENCY",
+                                                             str.c_str()));
+    }
+
+    if (!retval)
+    {
+        try {
+            auto guid{static_cast<GncGUID>(gnc::GUID::from_string(str))};
+            retval = qof_instance_from_guid(&guid, type);
+        }
+        catch (const gnc::guid_syntax_exception& err)
+        {
+            PWARN("Failed to convert %s to a GUID", str.c_str());
+        }
+    }
+    return retval;
+}
+
+std::string
+qof_instance_to_string(const QofInstance* inst)
+{
+    gnc::GUID guid{*qof_instance_get_guid(inst)};
+    return guid.to_string();
+}
+
+template <typename ValueType> void
+GncOptionValue<ValueType>::set_value(ValueType new_value)
+{
+    m_value = new_value;
+}
+
+template <typename ValueType> void
+GncOptionValue<ValueType>::set_default_value(ValueType new_value)
+{
+    m_value = m_default_value = new_value;
+}
+
+template <typename ValueType> void
+GncOptionValue<ValueType>::reset_default_value()
+{
+    m_value = m_default_value;
+}
+
+template <> void
+GncOptionValue<SCM>::set_value(SCM new_value)
+{
+    if (m_value)
+        scm_gc_unprotect_object(m_value);
+    m_value = new_value;
+    scm_gc_protect_object(m_value);
+}
+
+template <> void
+GncOptionValue<SCM>::set_default_value(SCM new_value)
+{
+    if (m_value)
+        scm_gc_unprotect_object(m_value);
+    if (m_default_value)
+        scm_gc_unprotect_object(m_default_value);
+    m_value = m_default_value = new_value;
+    scm_gc_protect_object(m_value);
+    scm_gc_protect_object(m_default_value);
+}
+
+template <> void
+GncOptionValue<SCM>::reset_default_value()
+{
+    if (m_value)
+        scm_gc_unprotect_object(m_value);
+    m_value = m_default_value;
+    scm_gc_protect_object(m_value);
+}
+
+template GncOptionValue<bool>::GncOptionValue(const GncOptionValue<bool>&);
+template GncOptionValue<int>::GncOptionValue(const GncOptionValue<int>&);
```

Comment:
Again for my enlightenment - what's the purpose of these templates ? Is that to force the compiler to generate all these variants of GncOpionalValue ?

Reply:
Exactly. Otherwise the template definitions have to be in headers which both slows down compilation and exposes implementation details of GncOption.

Reply:
Ok thanks.